### PR TITLE
Implement Runge-Kutta 4

### PR DESCRIPTION
**Pull request by axrolld**
Implement Runge-Kutta 4 method in OSRD.

### CODE REVIEW

File: src/main/java/fr/sncf/osrd/train/TrainPhysicsIntegrator.java [hunk 1]
```diff
@@ -87,6 +90,20 @@ public double computeAcceleration(Action action) {
         return (actionTractionForce + weightForce + effectiveOppositeForces) / inertia;
     }
 
+    /**
+     * Computes the action of the train, given a set of speedControllers
+     * @param controllers the set of speedControllers we're gonna get the action from
+     * @param end end position of the route
```
Comment (bgiuliana) [inline src/main/java/fr/sncf/osrd/train/TrainPhysicsIntegrator.java:100]: typo: end twice

File: src/main/java/fr/sncf/osrd/train/TrainPhysicsIntegrator.java [hunk 2]
```diff
@@ -197,87 +220,90 @@ public IntegrationStep stepFromAcceleration(double acceleration, double traction
...
-
-        var integrator = new TrainPhysicsIntegrator(
+    /**
+     * Compute the next integration step with runge-kutta 4 method
+     * @param initialLocation the location at the beginning of the step
+     * @param initialSpeed the speed at the beginning of the step
+     * @param rollingStock the rolling stock used for the step
+     * @param timeStep the time step used for the step
+     * @param end the end of the train's path
+     * @param directionSign +1 if forward circulation, -1 if backwards
```
Comment (bgiuliana) [inline src/main/java/fr/sncf/osrd/train/TrainPhysicsIntegrator.java:246]: you should update the comments with makeAction

File: src/test/java/fr/sncf/osrd/Helpers.java [hunk 3]
```diff
@@ -81,7 +81,7 @@ public static RJSSimulation loadExampleSimulationResource(Class<?> klass, String
         return sim;
     }
 
-    private static final boolean saveCSVFiles = false;
+    private static final boolean saveCSVFiles = true;
```
Comment (flomonster) [inline src/test/java/fr/sncf/osrd/Helpers.java]: This shouldn't be changed...

File: src/test/java/fr/sncf/osrd/speedcontroller/MarginTests.java [hunk 4]
```diff
@@ -227,7 +227,8 @@ public void testEcoMargin(double value, TestInfo info) {
     }
 
     @ParameterizedTest
-    @ValueSource(ints = {0, 1, 2, 3, 4, 5, 6, 7})
+    //@ValueSource(ints = {0, 1, 2, 3, 4, 5, 6, 7})
+    @ValueSource(ints = {0})
```
Comment (flomonster) [inline src/test/java/fr/sncf/osrd/speedcontroller/MarginTests.java]: Why reduce the number of tests?

**Review by flomonster — CHANGES_REQUESTED** [review]
I didn't get all the physics part but the pr seems good.

File: src/test/java/fr/sncf/osrd/train/TrainPhysics.java [hunk 5]
```diff
@@ -37,7 +60,7 @@ public void testSteepSlopeTraction() {
         double maxTraction = rollingStock.tractiveEffortCurve[0].maxEffort;
         // how fast would a train go after 10 steps of 1 sec, full throttle on a 45deg slope?
```
Comment (bgiuliana) [inline src/test/java/fr/sncf/osrd/train/TrainPhysics.java:61]: 45 deg does not correspond to the line 63?

File: src/test/java/fr/sncf/osrd/speedcontroller/MarginTests.java (hunk 4 above)
Comment (axrolld) [reply src/test/java/fr/sncf/osrd/speedcontroller/MarginTests.java]: My bad, i did it for debug but forgot to put it back to normal.

File: src/main/java/fr/sncf/osrd/train/TrainPhysicsIntegrator.java [hunk 6]
```diff
@@ -178,14 +199,16 @@ public IntegrationStep stepFromAction(Action action, double maxDistance, double
      * @param directionSign +1 if forward circulation, -1 if backwards
      * @return the new speed of the train
      */
-    public IntegrationStep stepFromAcceleration(double acceleration, double tractionForce,
-                                                              double maxDistance, double directionSign) {
+    public IntegrationStep stepFromAcceleration(double acceleration, double tractionForce, double maxDistance,
+                                                boolean isRKStep, double directionSign) {
```
Comment (bgiuliana) [inline src/main/java/fr/sncf/osrd/train/TrainPhysicsIntegrator.java:209]: comment isRKStep

File: src/main/java/fr/sncf/osrd/train/TrainPhysicsIntegrator.java [hunk 7]
```diff
@@ -178,14 +199,16 @@ public IntegrationStep stepFromAction(Action action, double maxDistance, double
...
 
         var newSpeed = currentSpeed + directionSign * acceleration * timeStep;
         var timeDelta = timeStep;
 
         // if the speed change sign or is very low we integrate only the step at which the speed is zero
-        if (currentSpeed != 0.0 && (Math.signum(newSpeed) != Math.signum(currentSpeed) || abs(newSpeed) < 1E-10)) {
+        if (!isRKStep && currentSpeed != 0.0
```
Comment (bgiuliana) [inline src/main/java/fr/sncf/osrd/train/TrainPhysicsIntegrator.java]: add a comment

File: src/test/java/fr/sncf/osrd/Helpers.java (hunk 3 above)
Comment (axrolld) [reply src/test/java/fr/sncf/osrd/Helpers.java]: I keep forgetting to set this back to false....

File: src/test/java/fr/sncf/osrd/train/TrainPhysics.java (hunk 5 above)
Comment (axrolld) [reply src/test/java/fr/sncf/osrd/train/TrainPhysics.java:61]: Yes is does, 45 deg is 1000 %o

File: src/main/java/fr/sncf/osrd/train/TrainPhysicsIntegrator.java (hunk 6 above)
Comment (axrolld) [reply src/main/java/fr/sncf/osrd/train/TrainPhysicsIntegrator.java:209]: Done

File: src/main/java/fr/sncf/osrd/train/TrainPhysicsIntegrator.java (hunk 7 above)
Comment (axrolld) [reply src/main/java/fr/sncf/osrd/train/TrainPhysicsIntegrator.java]: Done

**Review by bgiuliana — APPROVED** [review]

**Review by flomonster — APPROVED** [review]

**Pull request merged**